packages/ui/src/components/Passcode/index.tsx

```diff
@@ -0,0 +1,229 @@
+import classNames from 'classnames';
+import React, {
+  useMemo,
+  useRef,
+  useCallback,
+  FormEventHandler,
+  KeyboardEventHandler,
+  FocusEventHandler,
+  ClipboardEventHandler,
+} from 'react';
+
+import * as styles from './index.module.scss';
+
+export const defaultLength = 6;
+
+export type Props = {
+  name: string;
+  isDisabled?: boolean;
+  className?: string;
+  length?: number;
+  value: string[];
+  hasError?: boolean;
+  onChange: (value: string[]) => void;
+};
+
+const isNumeric = (char: string) => /^\d*$/.test(char);
+
+const normalize = (value: string[], length: number): string[] => {
+  if (value.length > length) {
+    return value.slice(0, length);
+  }
+
+  if (value.length < length) {
+    return value.concat(Array.from({ length: length - value.length }));
+  }
+
+  return value;
+};
+
+const trim = (oldValue: string | undefined, newValue: string) => {
+  // Trim oldValue from the latest input to get the updated char
+  if (newValue.length > 1 && oldValue) {
+    return newValue.replace(oldValue, '');
+  }
+
+  return newValue;
+};
+
+const Passcode = ({
+  name,
+  isDisabled,
+  className,
+  value,
+  length = 6,
+  hasError,
+  onChange,
+}: Props) => {
+  /* eslint-disable @typescript-eslint/ban-types */
+  const inputReferences = useRef<Array<HTMLInputElement | null>>(
+    Array.from<null>({ length }).fill(null)
+  );
+  /* eslint-enable @typescript-eslint/ban-types */
+
+  const codes = useMemo(() => normalize(value, length), [length, value]);
+
+  const onInputHandler: FormEventHandler<HTMLInputElement> = useCallback(
+    (event) => {
+      const { target } = event;
+
+      if (!(target instanceof HTMLInputElement)) {
+        return;
+      }
+
+      const { value, dataset } = target;
+
+      // Unrecognized target input field
+      if (dataset.id === undefined) {
+        return;
+      }
+
+      event.preventDefault();
+
+      // Filter non-numeric input
+      if (!isNumeric(value)) {
+        return;
+      }
+
+      const targetId = Number(dataset.id);
+
+      // Update the total input value
+      onChange(Object.assign([], codes, { [targetId]: trim(codes[targetId], value) }));
+
+      // Move to the next target
+      if (value) {
+        const nextTarget = inputReferences.current[targetId + 1];
+        nextTarget?.focus();
+        nextTarget?.select();
+      }
```

Comment:
what if it's already the last input? Will it be index out of range?

Reply:
yep, considering handling it on the upper scope.  Gonna have a container component wrap up the UI, will handle the code format validation, error display, count down, resend and submit logic. 
Might have a callback method props added later or use an onChange useEffect hook to handle that.  
 i.e. if the user updates a digit in the middle, should we `focus on the submit button ` or `auto submit` as well.

Reply:
ok